dspinner/pin.go

```diff
@@ -0,0 +1,1069 @@
+			tmpCidRIndex.Add(pp.cid.String(), pp.id)
+		} else if pp.mode == ipfspinner.Direct {
+			tmpCidDIndex.Add(pp.cid.String(), pp.id)
+		}
+		if pp.name != "" {
+			tmpNameIndex.Add(pp.name, pp.id)
+			hasNames = true
+		}
+	}
+
+	// Sync the CID index to what was build from pins.  This fixes any invalid
+	// indexes, which could happen if ipfs was terminated between writing pin
+	// and writing secondary index.
+	changed, err := p.cidRIndex.SyncTo(tmpCidRIndex)
+	if err != nil {
+		return fmt.Errorf("cannot sync indexes: %v", err)
+	}
+	if changed {
+		log.Info("invalid recursive indexes detected - rebuilt")
+	}
+
+	changed, err = p.cidDIndex.SyncTo(tmpCidDIndex)
+	if err != nil {
+		return fmt.Errorf("cannot sync indexes: %v", err)
+	}
+	if changed {
+		log.Info("invalid direct indexes detected - rebuilt")
+	}
+
+	if hasNames {
+		changed, err = p.nameIndex.SyncTo(tmpNameIndex)
+		if err != nil {
+			return fmt.Errorf("cannot sync name indexes: %v", err)
+		}
+		if changed {
+			log.Info("invalid name indexes detected - rebuilt")
+		}
+	}
+
+	return p.Flush(ctx)
+}
+
+// DirectKeys returns a slice containing the directly pinned keys
+func (p *pinner) DirectKeys(ctx context.Context) ([]cid.Cid, error) {
+	p.lock.RLock()
+	defer p.lock.RUnlock()
+
+	var cids []cid.Cid
+	var e error
+	err := p.cidDIndex.ForEach("", func(idx, id string) bool {
+		var c cid.Cid
+		c, e = cid.Decode(idx)
+		if e != nil {
+			return false
+		}
+		cids = append(cids, c)
+		return true
+	})
+	if err != nil {
+		return nil, err
+	}
+	if e != nil {
+		return nil, e
+	}
+
+	return cids, nil
+}
+
+// RecursiveKeys returns a slice containing the recursively pinned keys
+func (p *pinner) RecursiveKeys(ctx context.Context) ([]cid.Cid, error) {
+	p.lock.RLock()
+	defer p.lock.RUnlock()
+
+	var cids []cid.Cid
+	var e error
+	err := p.cidRIndex.ForEach("", func(idx, id string) bool {
+		var c cid.Cid
+		c, e = cid.Decode(idx)
+		if e != nil {
+			return false
+		}
+		cids = append(cids, c)
+		return true
+	})
+	if err != nil {
+		return nil, err
+	}
+	if e != nil {
+		return nil, e
+	}
+
+	return cids, nil
+}
+
+// InternalPins returns all cids kept pinned for the internal state of the
+// pinner
+func (p *pinner) InternalPins(ctx context.Context) ([]cid.Cid, error) {
+	return nil, nil
+}
+
+// Update updates a recursive pin from one cid to another.  This is equivalent
+// to pinning the new one and unpinning the old one.
+//
+// TODO: This will not work when multiple pins are supported
+func (p *pinner) Update(ctx context.Context, from, to cid.Cid, unpin bool) error {
+	p.lock.Lock()
+	defer p.lock.Unlock()
+
+	found, err := p.cidRIndex.HasAny(from.String())
+	if err != nil {
+		return err
+	}
+	if !found {
+		return fmt.Errorf("'from' cid was not recursively pinned already")
+	}
+
+	// If `from` already recusively pinned and `to` is the same, then all done
+	if from == to {
+		return nil
+	}
+
+	err = p.addPin(to, ipfspinner.Recursive, "")
```

Comment:
Fixed
There should have been a call to `dagutils.DiffEnumerate` that was missed.  It is more efficient because it fetches every object in the graph pointed to by `to` that is not in `from`.